### PR TITLE
Cql4bonnie qdm5 3 937 facility

### PR DESCRIPTION
**Pull request by losborne**
JIRA Test: https://jira.mitre.org/browse/BONNIE-968

Related PRs:
- https://github.com/projecttacoma/cql_qdm_patientapi/pull/72
- https://github.com/projecttacoma/bonnie/pull/819
- https://github.com/projectcypress/health-data-standards/pull/530

### CODE REVIEW

File: spec/javascripts/datatypes/encounter_spec.js.coffee [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  describe "Performed", ->
+    it "should show nothing if there is no facility", ->
+      encounterPerformed = new CQL_QDM.EncounterPerformed({})
+      expect(JSON.stringify(encounterPerformed.facilityLocations())).toEqual "[]"
+    
+    it "should show just the single facilityLocation", ->
+      encounterPerformed = new CQL_QDM.EncounterPerformed({'facility': {"type":"COL","values":[{"code":{"code_system":"HCPCS","code":"G0438"},"display":"Annual Wellness Visit","locationPeriodLow":"08/30/2017 1:00 AM","locationPeriodHigh":"08/31/2017 1:00 AM"}]}})
+      expect(JSON.stringify(encounterPerformed.facilityLocations())).toEqual '[{"code":{"code":"G0438","system":"HCPCS","display":"Annual Wellness Visit"},"locationPeriod":{"low":"2017-08-30T01:00:00.00+0000","high":"2017-08-31T01:00:00.00+0000","lowClosed":true,"highClosed":true}}]'
+
+    it "should show two Facilities", ->
```
Comment (edeyoung) [inline spec/javascripts/datatypes/encounter_spec.js.coffee]: can you add a test for when there's no facility location period? I'm wondering if the parsing code above will puke with no location period.

Comment (losborne) [inline spec/javascripts/datatypes/encounter_spec.js.coffee]: I added the test, expecting the current behavior, which shows this if there is no location period.  I'm not sure if that's the right behavior or not.
image

Comment (edeyoung) [inline spec/javascripts/datatypes/encounter_spec.js.coffee]: hm. I think it should be `nil` or `undefined` or something. glad it doesn't explode, though!

Comment (edeyoung) [inline spec/javascripts/datatypes/encounter_spec.js.coffee]: yeah, it definitely should be `nil` or `undefined`.

Comment (losborne) [inline spec/javascripts/datatypes/encounter_spec.js.coffee]: I've updated it to be undefined and added tests for no location period as well as location period with undefined start and end dates.

File: app/assets/javascripts/types/facility.js.coffee [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+Represents a CQL 'facility' for use in the CQL execution engine.
+###
+class CQL_QDM.Facility
+  constructor: (facility) ->
+    @code = new cql.Code(facility.code.code, facility.code.code_system, undefined, facility.display)
+
+    locationPeriodLow = CQL_QDM.Helpers.convertDateTime(facility.locationPeriodLow) if facility.locationPeriodLow
+    locationPeriodHigh = CQL_QDM.Helpers.convertDateTime(facility.locationPeriodHigh) if facility.locationPeriodHigh
+    # If only one of Low/High is defined, make an interval with an undefined start or end
+    @locationPeriod = new cql.Interval(locationPeriodLow, locationPeriodHigh) if locationPeriodLow || locationPeriodHigh
```
Comment (edeyoung) [inline app/assets/javascripts/types/facility.js.coffee]: If `locationPeriodHigh` exists, there must be a `locationPeriodLow`.

My intuition is that if `locationPeriodHigh` doesn't exist but `locationPeriodLow` does, the `high` for `locationPeriod` should be `null` (not a `null` for each intermediate value (month, day, etc.). Does this `if` statement do that appropriately? If not, is my intuition that this is the correct behavior wrong?

File: spec/javascripts/datatypes/encounter_spec.js.coffee [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      expect(JSON.stringify(encounterPerformed.facilityLocations())).toEqual '[{"code":{"code":"G0438","system":"HCPCS","display":"Annual Wellness Visit"},"locationPeriod":{"low":"2017-08-30T01:00:00.00+0000","high":"2017-08-31T01:00:00.00+0000","lowClosed":true,"highClosed":true}},{"code":{"code":"105401000119101","system":"SNOMED-CT","display":"Diabetes"},"locationPeriod":{"low":"2017-08-30T01:00:00.00+0000","high":"2017-08-31T01:00:00.00+0000","lowClosed":true,"highClosed":true}}]'
+
+    it "should show Facility without locationPeriod if none provided", ->
+      encounterPerformed = new CQL_QDM.EncounterPerformed({'facility': {"type":"COL","values":[{"code":{"code_system":"HCPCS","code":"G0438"},"display":"Annual Wellness Visit","locationPeriodLow":null,"locationPeriodHigh":null}]}})
+      expect(JSON.stringify(encounterPerformed.facilityLocations())).toEqual '[{"code":{"code":"G0438","system":"HCPCS","display":"Annual Wellness Visit"}}]'
+
+    it "should show locationPeriod with undefined start date", ->
```
Comment (edeyoung) [inline spec/javascripts/datatypes/encounter_spec.js.coffee]: I think the title for this test and the `undefined end date` test got mixed up.

**Pull request merged**